vm-virtio/src/vhost_user/handler.rs

```diff
@@ -0,0 +1,200 @@
+    pub fn run(&mut self) -> Result<()> {
+        let epoll_fd = epoll::create(true).map_err(Error::EpollCreateFd)?;
+
+        for (index, vhost_user_interrupt) in self.vu_interrupt_list.iter().enumerate() {
+            epoll::ctl(
+                epoll_fd,
+                epoll::ControlOptions::EPOLL_CTL_ADD,
+                vhost_user_interrupt.as_raw_fd(),
+                epoll::Event::new(epoll::Events::EPOLLIN, index as u64),
+            )
+            .map_err(Error::EpollCtl)?;
+        }
+
+        let _cvq_event = self.vu_interrupt_list.len();
+        match &self.cvq {
+            Some(cvq) => {
+                epoll::ctl(
+                    epoll_fd,
+                    epoll::ControlOptions::EPOLL_CTL_ADD,
+                    cvq.queue_evt.as_raw_fd(),
+                    epoll::Event::new(epoll::Events::EPOLLIN, _cvq_event as u64),
+                )
+                .map_err(Error::EpollCtl)?;
+            }
+            None => println!("no ctrl queue event to handle!\n"),
+        }
+
+        let _kill_event = _cvq_event + 1;
+        epoll::ctl(
+            epoll_fd,
+            epoll::ControlOptions::EPOLL_CTL_ADD,
+            self.kill_evt.as_raw_fd(),
+            epoll::Event::new(epoll::Events::EPOLLIN, _kill_event as u64),
+        )
+        .map_err(Error::EpollCtl)?;
+
+        let mut events = vec![epoll::Event::new(epoll::Events::empty(), 0); _kill_event];
+
+        loop {
+            let num_events =
+                epoll::wait(epoll_fd, -1, &mut events[..]).map_err(Error::EpollWait)?;
+
+            for event in events.iter().take(num_events) {
+                let index = event.data as usize;
+                if index < _cvq_event {
+                    let vhost_user_interrupt = &self.vu_interrupt_list[index];
+                    vhost_user_interrupt
+                        .read()
+                        .map_err(Error::FailedReadingQueue)?;
+                    let result = self.signal_used_queue(&self.queues[index]);
+                    if let Err(_e) = result {
+                        error!("failed to signal used queue");
+                    }
+                } else if index == _cvq_event {
+                    println!("control queue event received!\n");
+                    if let Some(cvq) = self.cvq.take() {
+                        if let Err(_e) = cvq.queue_evt.read() {
+                            error!("failed to get ctl queue event: {:?}", _e);
+                        }
+                        if let Err(_e) = self.process_cvq(cvq) {
+                            error!("failed to process ctl queue: {:?}", _e);
+                        }
+                    } else {
+                        error!("No control queue info !\n");
+                    }
+                } else if index == _kill_event {
+                    info!("vhost device removed");
+                    return Ok(());
+                } else {
+                    error!("Unknow event for vhost-net");
```

Comment:
Typo.

Reply:
Yes, right, typo! Thanks for pointing out this, I will correct it in next version.

Reply:
You spelt unknown wrong. But it should also be vhost-user-net.

Reply:
Change is done!